SIPS/sip-5.md

```diff
@@ -0,0 +1,188 @@
+---
+sip: 5
+title: Creating JSON Web Tokens (JWTs)
+status: Draft
+discussions-to: https://github.com/MetaMask/SIPs/discussions/64
+author: Vid Kersic <vid.kersic@yahoo.com> (@Vid201), Andraz Vrecko (@andyv09)
+created: 2022-10-20
+---
+
+## Abstract
+
+This SIP proposes a functionality of signing the data in the form of JSON Web Tokens (JWTs), exposed as an RPC method.
+
+JSON Web Tokens (JWTs) are an internet standard for creating data whose payload holds claims in the JSON. The standard is widely used in many applications on the web, often for authentication and authorization purposes. Other use cases are Verifiable Credentials (VCs) and Verifiable Presentations (VPs), also part of Decentralized Identity and Self-Sovereign Identity (SSI).
+
+## Motivation
+
+Snaps were introduced to extend the functionalities of MetaMask to support more use cases and paradigms. MetaMask currently provides several RPC methods for digitally signing data (described [here](https://docs.metamask.io/guide/signing-data.html)), such as personal_sign and signTypedData. All methods add the prefix "\x19Ethereum Signed Message:\n" to the data, as described in [EIP 191](https://eips.ethereum.org/EIPS/eip-191) and [EIP 712](https://eips.ethereum.org/EIPS/eip-712), which prevents several attack vectors, most notably impersonating transactions. But this also prevents the ability to produce pure signatures over the data, which is needed for other internet data standards (such as JWTs). Actually, a pure signature is possible only with eth_sign, but that method was deprecated and advised not to use. Therefore, there is no safe way to create JWTs containing signatures signed by MetaMask accounts. This SIP proposes a new RPC method for Snaps that enables the creation of signed JWTs.
```

Comment:
Can you explain the rationale of why the snap itself can't calculate JWT cookies using `snap_getBip44Entropy`?

The above function gives you the raw, extended, private key that is used inside MetaMask itself. With that entropy you can do calculations inside of the snap, without needing additional support from the extension.

Reply:
It is possible. There are two ways to do it without additional support:

1. Using the RPC method `eth_sign`, but this method is deprecated, and the user can see a red notice when signing.
2. Exporting private key with `snap_getBip44Entropy`, but then the user must confirm that it allows Snap to export private key (when installing Snap) from "main" MetaMask to the Snap. But this would look suspicious to users and prevent them from using the Snap. We also need to sign with the same private key as used on the Ethereum, which means Snap has complete access to the funds on the mainnet.

Understandably, MetaMask and Snap JSON-RPC API cannot support all signing methods, but JWTs are one of the most used standards on the internet, and many Snaps could reuse this functionality.